semantics-3.0/Semantics.hs

```diff
@@ -417,9 +417,14 @@ addOutcome party diffValue trOut = let
 
 -- | Extract total outcomes from transaction inputs and outputs
 getOutcomes :: [Payment] -> [Input] -> TransactionOutcomes
-getOutcomes effect input = let
-    outcomes = [ (party, money) | Payment party money <- effect ]
-    incomes  = [ (party, money) | IDeposit _ party money <- input ]
+getOutcomes payments input = let
+    outcomes = map (\(Payment party money) -> (party, money)) payments
+
+    deposits (IDeposit _ party money) acc = (party, money) : acc
+    deposits _ acc = acc
+
+    incomes = foldr deposits [] input
+
```

Comment:
Come on, which one is easier to read? If you really don't like list comprehensions then I would use `mapMaybe` instead of `fold`. Actually that could be a good alternative for PS, mapMaybe + maybe monad.

Reply:
Oh wait, that is what David did already, good thinking :)

Reply:
I like this `foldr`, to me it's just as easy as `mapMaybe`. I'm not used to `mapMaybe` so I don't instantly know what it does therefore `foldr` doing basically a specialised version of the same thing works well.

Reply:
The list comprehension has been in mathematics for a while though, and is much more concise. Just my opinion anyway, in Isabelle I've used simple recursion.

Reply:
I never liked list comprehensions, to me they are just a specialised version of do notation with extra magic but then I haven't used them in maths for about 15 years 😃 . Inline `do` notation is almost as close to set builder notation anyway since it has the curly braces :D `do { pure x; x <- xs }`

Reply:
Exactly, that is why I was suggesting the mapMaybe + maybe monad, they are pretty close. I never thought I would be siding for mathematical notation, I must be getting old. But when I learned about list comprehension (with Erlang) I thought it was so convenient, if only there was something similar that was as powerful as fold... We could get rid of fold all together :)

Reply:
The issue in this case though is the magic pattern matching filter stuff

Reply:
Ah, so maybe not with `mapMaybe`, but can you not do something like:
```haskell
outcomes = do { Payment party money <- effect;
                return (party, money) }
```

Reply:
It seems to work in Haskell, and I imagine it works in PureScript with arrays, but not sure with lists.

Reply:
Actually, I like list comprehension a lot :-)
But I agree with David here, that the implicit filtering is magic, and should be stated explicitly.
The main reason I chose `foldr` is, again, to use (kind of) "well-known" functions.